include/ekuvector/ekuvector.hpp

```diff
@@ -0,0 +1,1122 @@
+  other.data_ = nullptr;
+}
+
+template <class Type, class Allocator>
+ekuvector<Type, Allocator>::ekuvector(ekuvector &&other, const Allocator &alloc)
+    : ekuvector(alloc) {
+  /* the rest of the move operation changes if source and
+     destination have equivalent allocators */
+  auto other_alloc = std::allocator_traits<allocator_type>::
+      select_on_container_copy_construction(other.get_allocator());
+
+  if (alloc == other_alloc) {
+    // relocate assets from the source to the destination
+    capacity_ = other.capacity_;
+    size_ = other.size_;
+    data_ = other.data_;
+    // empty source and leave in a safe state
+    other.size_ = 0;
+    other.capacity_ = 0;
+    other.data_ = nullptr;
+  } else {
+    // preallocate a new memory block and move construct the elements in the
+    // original container into the destination
+    preallocate_capacity(other.size());
+    // move-construct items at the end of this list
+    while (size_ < other.size_) {
+      emplace_back(std::move(other[size_]));
+    }
+  }
+}
+
+template <class Type, class Allocator>
+ekuvector<Type, Allocator>::ekuvector(std::initializer_list<Type> init,
+                                      const Allocator &alloc)
+    : ekuvector(alloc) {
+  /* preallocate memory space */
+  preallocate_capacity(init.size());
+  /* copy constructo elements at the end of this vector */
+  for (auto it = init.begin(); it != init.end(); ++it) {
+    emplace_back(*it);
+  }
+}
+
+template <class Type, class Allocator>
+ekuvector<Type, Allocator>::~ekuvector() {
+  /* make sure all destructors get called before I release the memory block */
+  clear();
+  /* release memory */
+  if (capacity_) {
+    allocator_.deallocate(data_, capacity_);
+    capacity_ = 0;
+  }
+}
+
+template <class Type, class Allocator>
+void ekuvector<Type, Allocator>::reserve(size_type new_cap) {
+  /* if the new capacity is smaller than the current size, don't do anything */
+  if (new_cap <= capacity_) {
+    return;
+  }
+
+  auto new_data_ptr = allocator_.allocate(new_cap);
+  auto new_capacity = new_cap;
+
+  for (size_t index = 0; index < size_; ++index) {
+    allocator_.construct(new_data_ptr + index, std::move(*(data_ + index)));
+  }
+
+  /* destroy the empty shells left in the old buffer, and then release it */
+  clear();
+  if (capacity_) {
+    allocator_.deallocate(data_, capacity_);
+  }
+
+  /* replace with the new block */
+  data_ = new_data_ptr;
+  capacity_ = new_capacity;
+}
+
+template <class Type, class Allocator>
+void ekuvector<Type, Allocator>::preallocate_capacity(size_type new_cap) {
+  /* if the new capacity is smaller than the current size, don't do anything */
+  auto rounded_size = ((new_cap + 1023) / 1024) * 1024;
+  reserve(rounded_size);
+}
+
+template <class Type, class Allocator>
+ekuvector<Type, Allocator> &ekuvector<Type, Allocator>::
+operator=(const ekuvector &other) {
+  /* clear up the current contents */
+  clear();
+  preallocate_capacity(other.size());
```

Comment:
@glpuga is this code exception safe? What would happen if `preallocate_capacity` throws? Same everywhere else `allocator_` is used.

Reply:
I have not really studied the exception safety of this implementation in detail. However, if 'preallocate_capacity()' throws, depending on what member function this happens the exception guarantee is either strong (constructors, assign for single element, insert, push_back, and other members that start by resizing the buffer first), or basic (operator=, assign with input iterators, and others that do some irreversible, but stable, stuff before resizing).

Some of these can be improved. 'operator=', for instance, can be made to have a strong exception guarantee for memory allocation exceptions by trivially changing the order in which 'clear()' and 'preallocate_capacity()' get called, though this trivial improvement may potentially impact performance by causing preallocate to transfer the vector to a new buffer right before emptying it whenever a reallocation is needed. Better implementations than this can be found.

If datatype constructors or copy/move operations throw, it's possible that the implementation provides no exception guarantee.